### PR TITLE
Geojson and  Table style changes

### LINKED ISSUE #6956: Add `czmlTemplate` support for `LineString/MultiLineString`
This would allow us to use the following CZML objects

- https://github.com/AnalyticalGraphicsInc/czml-writer/wiki/Polyline
- https://github.com/AnalyticalGraphicsInc/czml-writer/wiki/PolylineVolume
- https://github.com/AnalyticalGraphicsInc/czml-writer/wiki/Wall
- https://github.com/AnalyticalGraphicsInc/czml-writer/wiki/Corridor

### PR DESCRIPTION
**Pull request by na9da**
### What this PR does

Fixes https://github.com/TerriaJS/terriajs/issues/6956 https://github.com/TerriaJS/terriajs/issues/6374

This PR makes the following changes:
- Add support for multiple `urls` for `GeoJsonCatalogItem` ([commit](https://github.com/TerriaJS/terriajs/pull/7002/commits/a5c7269f62741392b35922b081eea720e372d171))
- Automatically explode GeoJSON `MultiPoint` features to `Point` features ([commit](https://github.com/TerriaJS/terriajs/pull/7002/commits/0a121eabc3fca080334642e7951a4241a2fabd00))
- Add new table styling traits - `scaleByDistance` and `disableDepthTestDistance` ([commit](https://github.com/TerriaJS/terriajs/pull/7002/commits/4ee5b43e49822e9666898656407b9f3c0308ef6e))
- Add support for `LineString` and `MultiLineString` when using `GeoJsonCatalogItem` in `CZML` mode. ([commit](https://github.com/TerriaJS/terriajs/pull/7002/commits/230e489486fc740820d7953226e711232bbf4ac1))

### Test me

This [test link](http://ci.terria.io/geojson-table-changes/#share=s-cVNATqHXLcE5E14CWgwYBSg87QE&https://gist.githubusercontent.com/na9da/c0bdef76799379795036ea2bb5336950/raw/bd5394c8131522c4246596d2e4cee1c35b4658d0/scale-by-distance.json) demonstrates all the above changes (except the LineString one). The catalog definition for the test link defines [multiple URLs](https://gist.github.com/na9da/c0bdef76799379795036ea2bb5336950#file-scale-by-distance-json-L6-L13), and the new traits [scaleByDistance and disableDepthTestDistance](https://gist.github.com/na9da/c0bdef76799379795036ea2bb5336950#file-scale-by-distance-json-L21-L27). The URL endpoints also return `MultiPoint` GeoJSON features which can now be correctly styled using table styling.

Notice how:
- Responses from 2 different endpoints are merged together as a single GeoJSON catalog item
- MultiPoint features are correctly styled using table styling properties
- the points are rendered small when zoomed out and scale up when zooming in. This reduces visual noise.
- add the Label and Marker items from the catalog, to see a similar effect.

### Checklist

- [x] There are unit tests to verify my changes are correct or unit tests aren't applicable (if so, write quick reason why unit tests don't exist)
- [ ] I've updated relevant documentation in `doc/`.
- [x] I've updated CHANGES.md with what I changed.
- [x] I've provided instructions in the PR description on how to test this PR.

### CODE REVIEW

File: lib/Models/Catalog/CatalogItems/GeoJsonCatalogItem.ts [hunk 1]
```diff
@@ -46,6 +51,45 @@ class GeoJsonCatalogItem
     return isDefined(this._file);
   }
 
+  /**
+   * Tries to build a FeatureCollection from partial geojson data
+   */
+  private featureCollectionFromPartialData(
+    jsonData: JsonValue
+  ): FeatureCollectionWithCrs | undefined {
```
Comment (na9da) [inline lib/Models/Catalog/CatalogItems/GeoJsonCatalogItem.ts:59]: Extracted out previous code that builds a full `FeatureCollection` from possibly partial API responses as we have to now use it in a couple of places.

File: lib/Models/Catalog/CatalogItems/GeoJsonCatalogItem.ts [hunk 2]
```diff
@@ -67,73 +111,91 @@ class GeoJsonCatalogItem
         jsonData = await readJson(this._file);
       }
     }
+    // We have multiple sources.
+    else if (this.urls.length > 0) {
```
Comment (na9da) [inline lib/Models/Catalog/CatalogItems/GeoJsonCatalogItem.ts:115]: Handle multi URL configuration:
- Fetch URL
- Build `FeatureCollection` from response
- Reproject `FeatureCollection`
- Merge all `FeatureCollection`s

File: lib/Models/Catalog/CatalogItems/GeoJsonCatalogItem.ts [hunk 3]
```diff
@@ -67,73 +111,91 @@ class GeoJsonCatalogItem
...
-          }
+    throw TerriaError.from(
+      "Invalid geojson data - only FeatureCollection and Feature are supported"
+    );
+  }
 
-          if (typeof geojson === "string") {
-            geojson = JSON.parse(geojson);
-          }
+  private async fetchSource(
```
Comment (na9da) [inline lib/Models/Catalog/CatalogItems/GeoJsonCatalogItem.ts:160]: Extracted out of `forceLoadGeoJsonData`

File: lib/Models/Catalog/CatalogItems/GeoJsonCatalogItem.ts [hunk 4]
```diff
@@ -67,73 +111,91 @@ class GeoJsonCatalogItem
...
+    return jsonData;
   }
 }
 
+function mergeFeatureCollections(
+  featureCollections: Array<FeatureCollection>
+): JsonValue {
+  return featureCollection(
+    featureCollections.map((fc) => fc.features).flat()
+  ) as any as JsonValue;
```
Comment (nf-s) [inline lib/Models/Catalog/CatalogItems/GeoJsonCatalogItem.ts]: can this be typed as `FeatureCollection`?

**Review by nf-s — CHANGES_REQUESTED** [review]
Looks good Nanda. Just a few small comments

- Can you please add a quick test for `disableDepthTestDistance`
  - You should be able to tack it on to `test/ModelMixins/TableMixinSpec.ts`
- It's not essential, but it would be nice to add these two new properties to `TableStylingWorkflow`

File: lib/ModelMixins/GeojsonMixin.ts [hunk 5]
```diff
@@ -948,6 +964,38 @@ function GeoJsonMixin<T extends AbstractConstructor<BaseType>>(Base: T) {
...
+            const positions: number[] = [];
+
+            geom.forEach((coords) => {
+              if (isJsonNumber(this.czmlTemplate?.heightOffset)) {
+                coords[2] = (coords[2] ?? 0) + this.czmlTemplate!.heightOffset;
+              }
+              positions.push(coords[0], coords[1], coords[2]);
+            });
+
+            czml.polyline.positions = { cartographicDegrees: positions };
```
Comment (nf-s) [inline lib/ModelMixins/GeojsonMixin.ts]: Can you please extend this to also support `czml.corridor`, `czml.wall` and `czml.polylineVolume`?

They all support the `positions: PositionList` property 


- https://github.com/AnalyticalGraphicsInc/czml-writer/wiki/Polyline
- https://github.com/AnalyticalGraphicsInc/czml-writer/wiki/PolylineVolume
- https://github.com/AnalyticalGraphicsInc/czml-writer/wiki/Wall
- https://github.com/AnalyticalGraphicsInc/czml-writer/wiki/Corridor

Adding support for these makes `czmlTemplate` fully feature-complete!

You don't need to validate the CZML - I would just check to see if `czml.polyline`, `czml.corridor`, `czml.wall` or `czml.polylineVolume` are defined - and if so, then add `positions` to each of them.

**Review by nf-s — CHANGES_REQUESTED** [review]

File: lib/Models/Catalog/CatalogItems/GeoJsonCatalogItem.ts (hunk 4 above)
Comment (na9da) [reply lib/Models/Catalog/CatalogItems/GeoJsonCatalogItem.ts]: I have changed the type, but note that we still need to do a forced casting of the `FeatureCollection` to `JsonValue`. For some reason TS doesn't like the type.

File: lib/ModelMixins/GeojsonMixin.ts (hunk 5 above)
Comment (na9da) [reply lib/ModelMixins/GeojsonMixin.ts]: Done, but I haven't tested them.

**Comment by na9da** [comment]
Have fixed most except the `TableStylingWorkflow` change which could be a separate issue?

**Review by nf-s — APPROVED** [review]
For some reason there is non-related test failing 🤔 - apart from that this looks great, thanks @na9da !

**Pull request merged**